### DISCUSSION TITLE
Speeding up MIMSunit metric for use in GGIR package

### DISCUSSION THREAD

**Discussion by vincentvanhees** [mHealth-Research-Group/MIMSunit#36] [Ideas]
Hi, I would like to embed your MIMS metric in the R package [GGIR](https://github.com/wadpac/GGIR) that I maintain. GGIR facilitates a variety of acceleration metrics. Adding the MIMS metric seems logical given its increased popularity. 

I noticed that MIMSunits computes substantially slower than the other metrics GGIR currently facilitates. By the way, I am using the CRAN version of MIMSunit, not sure if that matters. This led me to implement a lighter version of MIMS that is at least 20 times faster with only marginal value differences compared with the original. I achieved this by rewriting the code and stripping out all the parts that I think are not really critical for physical activity research. Please find my code and a comparison of the metrics below, and additional notes on differences at the bottom of this post.

**My plan:**
My plan is to add this 'light' version of MIMS to GGIR in it's next release, with reference to your paper and with clear guidance to the GGIR-user that my implementation is not an exact replication of your MIMS but practically identical with reference to the results below. 

**Why this issue?**
I am posting this issue here just in case you see any potential source of concern about the approach I am taking.

## Function to calculate my 'light' version of MIMSunit:

```
library(signal)

MIMSlight = function(df, epochsize=5, acc_columns = 2:4) {
  averageperepoch = function(x,sf,epochsize) { 
    # I am aware that this is numerically unstable, but in our use-case
    # expected drop in precision is far beyond required precision
    # and by that acceptable.
    x2 =cumsum(c(0,x))
    select = seq(1,length(x2),by=sf*epochsize)
    x3 = diff(x2[round(select)]) / abs(diff(round(select)))
  }
  bf_filter = function(lb, hb, n, sf) {
    hb = ifelse(test = sf <= (hb*2), yes = (sf / 2) - 0.1, no = hb)
    Wc = matrix(0,2,1)
    Wc[1,1] = lb / (sf/2) # lb: lower boundary
    Wc[2,1] = hb / (sf/2) # hb: higher boundary
    return(signal::butter(n,Wc,type=c("pass"))) 
  }
  mims_per_axis = matrix(NA, floor(nrow(df) / (sf*epochsize)), 3)
  for (i in acc_columns) {
    coef = bf_filter(lb=0.2, hb=5, n=4, sf=sf)
    filtered_signal = abs(signal::filter(coef, df[,i]))
    BFSM_per_axis = averageperepoch(filtered_signal, sf, epochsize)
    noise = which(BFSM_per_axis < 0.01) # MIMSunit gives zeros when sensor not moving, so we do same
    if (length(noise) > 0) BFSM_per_axis[noise] = 0
    mims_per_axis[,i-1] = BFSM_per_axis * epochsize
    
  }
  MIMSlight = rowSums(mims_per_axis)
  return(MIMSlight)
}
```

## Load data

I am loading 24 hours of raw data out of a multi-day ActiGraph recording which had sleep.mode disabled. I cannot share the file, but trust that you have a similar file lying around.

```
 library(MIMSunit); library(tibble); library(data.table); library(irr)
Nhour = 24 # number of hours of data to select
ActigraphFile = "D:/testfile.csv.gz"
sf = 100
rawacc = data.frame(data.table::fread(file = ActigraphFile, skip=36*3600*sf, nrow=Nhour*3600*sf, dec=","))
colnames(rawacc) = c("X","Y","Z")
now = Sys.time() # add dummy timestamps, we do not care what the actual time was for this demo.
HEADER_TIME_STAMP =  seq(now, (now+Nhour*3600)-(1/sf), by = 1/sf)
options(digits.secs = 3)
rawacc$HEADER_TIME_STAMP = as.POSIXct(HEADER_TIME_STAMP, origin="1970-1-1", tz = "")
rawacc = rawacc[, c("HEADER_TIME_STAMP", "X","Y","Z")] # order of columns is critical
rawacc = as_tibble(rawacc)
print(head(rawacc))

# A tibble: 6 x 4
  HEADER_TIME_STAMP            X      Y      Z
  <dttm>                   <dbl>  <dbl>  <dbl>
1 2021-07-18 12:08:45.412 -0.041 -0.545 -0.806
2 2021-07-18 12:08:45.422 -0.044 -0.548 -0.804
3 2021-07-18 12:08:45.432 -0.041 -0.545 -0.804
4 2021-07-18 12:08:45.442 -0.041 -0.545 -0.804
5 2021-07-18 12:08:45.452 -0.041 -0.545 -0.806
6 2021-07-18 12:08:45.462 -0.041 -0.545 -0.804
````
## Apply both metrics to the data

**Original MIMSunit:**

```
MIMSunit_timer_start = Sys.time()
S = MIMSunit::mims_unit(df = rawacc, epoch = '5 sec',
                        dynamic_range = c(-8, 8),
                        output_mims_per_axis = TRUE)

MIMSunit_timer_end = Sys.time()
speed_MIMSunit = difftime(MIMSunit_timer_end,MIMSunit_timer_start,units = "secs")
```

**MIMSlight (my lighter version of MIMS)**

```
rawacc = as.data.frame(rawacc) # removing tibble structure as it serves no purpose for light calculation
MIMSlight_timer_start = Sys.time()
MIMS = MIMSlight(df=rawacc, epochsize=5, acc_columns = 2:4)
MIMSlight_timer_end = Sys.time()
speed_MIMSlight = difftime(MIMSlight_timer_end, MIMSlight_timer_start, units="secs")

cat(paste0("\nRelative speed increase: ", as.numeric(speed_MIMSunit) / as.numeric(speed_MIMSlight),"\n"))

Relative speed increase: 24.36
```

## Visual comparison

```
# Truncate matrix, because S tends to be an epoch longer (I haven't investigate yet why)
if (nrow(S) > length(MIMS)) S = S[1:length(MIMS),]
png(filename = "D:/mimsplot.png",width = 9, height = 9, units = "in", res = 300)
par(mfrow=c(3,1))
plot(MIMS, S$MIMS_UNIT, type="p", pch=20, xlab="MIMSlight", ylab="MIMSoriginal", bty="l")
plot(S$MIMS_UNIT, MIMS - S$MIMS_UNIT, type="p", pch=20, xlab="MIMSoriginal", ylab="MIMSlight - MIMSoriginal", bty="l")
plot(MIMS, type="l", xlab="index", ylab="MIMSunit", bty="l")
lines(S$MIMS_UNIT, type="l", col="red")
legend("topright", legend = c("MIMSlight","MIMSoriginal"), col=c("black", "red"), lty=c(1,1))
dev.off()
```

which results in:
![mimsplot](https://user-images.githubusercontent.com/8362333/126065329-fe5f18e5-aa16-4a4b-af5d-2853e3c7b90d.png)

## Quantitative comparison

```
print(t.test(MIMS, S$MIMS_UNIT))
Welch Two Sample t-test
data:  MIMS and S$MIMS_UNIT
t = 0.085407, df = 34558, p-value = 0.9319
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.01381059  0.01506900
sample estimates:
mean of x mean of y 
0.3037958 0.3031666 

print(cor(MIMS, S$MIMS_UNIT))
[1] 0.9981458

print(irr::icc(cbind(MIMS, S$MIMS_UNIT), model = "twoway", 
  type = "agreement", unit = "single"))
 Single Score Intraclass Correlation
   Model: twoway 
   Type : agreement 
   Subjects = 17280 
     Raters = 2 
   ICC(A,1) = 0.998
 F-Test, H0: r0 = 0 ; H1: r0 > 0 
F(17279,17271) = 1076 , p = 0 
 95%-Confidence Interval for ICC Population Values:
  0.998 < ICC < 0.998
```

### Some additional notes:
- I suppose more work can be done to improve the resemblance. For example, the choice of Butterworth implementation, the alignment of samples and epochs in the two algorithms, and the noise thresholds I added.
-	It is not clear to me why you combined the three orthogonal axes by taking the sum instead of the geometric mean. There is clear geometric info in  the three axes, which would justify taking a geometric mean. Nevertheless, to preserve comparability, I copied your summation approach.
-	I stripped out the ActiGraph sleep-mode handling. I consider this a sensor brand specific issue and not something that should be dealt with inside the calculation of an acceleration metric.
-	I stripped out the timestamps and assume that any inaccuracy in sampling is either taken care of before metric calculation happens or is assumed not critical for research use-case.
-	As you can see your calculation of the area under the curve can be replaced by the calculation of the mean and multiplying it by epoch length, because it is an unsigned signal, so volume and mean only differ by a constant: the epoch length.
-	I have called my metric MIMSlight here to ease the narrative. However, I am not planning to call it MIMSlight once implemented in GGIR. Instead I will calculate BFMN (Band-pass Filtered, rectified, average per epoch, followed by MeaN of three axes) in _g_-units. Additionally, I will complement this with guidance to the user on how to convert this to MIMSunits: Multiply by three and then multiply that by the epoch length in seconds. I will do it this way, because I prefer to express data as much as possibly in _g_-units as that aids validation against ground truth measures of acceleration. Further, I prefer to avoid having a performance claim in the naming of the metrics GGIR facilitates.

**Comment by intille** [comment]
Hello, Vincent. 

Thanks for your note and we appreciate you taking a close look at the code. We implemented the algorithm the way we did (e.g., using relatively computationally expensive extrapolation) to make it more viable to compare across sensors with different specs which hadn’t been done before.   

We think it is great that you are looking for ways to develop algorithms that work more efficiently with particular types of data, but since you are creating a new algorithm by removing extrapolation, changing thresholds, etc., we don’t think this should be associated with the original work that was published and used in NHANES. The reason we proposed MIMS-Units was to create a metric that will work across a wide range of devices that are not limited to just +/- 8g sensors. We think that it’s important that people understand if there are changes to the algorithm, the conversion across devices may not work unless that is extensively tested. 

By saying that algorithm “is not an exact replication of […] MIMS but practically identical with reference to the results below,” you imply that it will work similarly. This could cause a lot of confusion in the community and so we do not recommend this. We do not think without replicating the type of cross-device testing done in our original paper, an algorithm should be stated to be, or implied to be, a faster/lighter version.

**Comment by vincentvanhees** [comment]
Hello Stephen
Thanks, I see your point. Causing more confusion is indeed not my intention.
I agree that it is best for me not to release this feature until I am able to accompany it with robust evidence to guide the user. 

Moving forward I will:
- Assess whether re-inserting your extrapolation gives a more accurate replication of MIMSunit and, if yes, whether there is still a speed advantage. => Probably not, but I will report back if there is.
- My student and I did our own mechanical shaker experiments with a variety of sensor brands, including some of the brands you did not test. I plan to perform a systematic comparison of metric speed and degree of monitor independence under a variety of device configuration conditions. => If we end up discovering anything disturbing about MIMSunit I will let you.

**Comment by vincentvanhees** [comment]
Update: Inserting your extrapolation function and replacing my band-pass filter with your filter hardly improves the agreement between our approaches. However, when I use your instead of my own aggregation approach the discrepancies almost entirely disappear.

In other words, when I use:

```
MIMSlight = function(df, epochsize=epochsize, acc_columns = 2:4, sf) {
  bf_filter = function(lb, hb, n, sf) {
    hb = ifelse(test = sf <= (hb*2), yes = (sf / 2) - 0.1, no = hb)
    Wc = matrix(0,2,1)
    Wc[1,1] = lb / (sf/2) # lb: lower boundary
    Wc[2,1] = hb / (sf/2) # hb: higher boundary
    return(signal::butter(n,Wc,type=c("pass")))
  }
  for (i in acc_columns) {
    coef = bf_filter(lb=0.2, hb=5, n=4, sf=sf)
    df[,i] = signal::filter(coef, df[,i])
  }
  integrated_data = MIMSunit::aggregate_for_mims(df,
                                                 epoch=paste0(epochsize,' sec'),
                                                 method="trapz",
                                                 rectify=TRUE,
                                                 st=NULL)
  MIMSlight = rowSums(integrated_data[,acc_columns])
  return(MIMSlight)
}
```

I get:

![mimsplot](https://user-images.githubusercontent.com/8362333/128502644-c6a1b5a5-8774-401e-ac11-7d0a12bb9113.png)

This function now runs three times faster than MIMSunit, so much slower than before (24 time faster) but with much better agreement.

The disagreement between our methods near zero (middle plot, left) may be explained by the fact that I am not doing the `.extrapolate_mark` step in your code... I haven't tried to replicate that yet.

None of this takes away the need for further research, but at least it seems useful to know where the differences in speed and output value come from.

PS. I closed this issue as I guess this is not really a direct issue with your code but more of an exchange of observations.

**Comment by qutang** [comment]
I converted the issue to the discussion here. So that we can keep this visible and open.

**Comment by dsherman11** [comment]
Time to revive this discussion, as I have further questions about using MIMS with GGIR.  Like Vincent, I also have questions about whether subtle changes to the MIMS algorithms are allowed.  

My company is developing a wearable device for scientific studies.  We would like to offer MIMS-units as one of our calculated metrics.

Regarding interpolation:  We are already sampling at 100Hz, so interpolation is not required.  We can clearly skip this step.

Regarding extrapolation:  Our sensor auto-ranges, and the signal encodes +/-16g with 15-bit resolution.  Clipping at this level will be rare.  Two questions:  

1. Is extrapolation still necessary when raw data is acquired at +/-16g?  
2. If we use a simpler extrapolation algorithm that better suits our device, are the results still MIMS-units?

Regarding aggregation:  In each of Vincent's modifications above, a single value is returned for each epoch, representing the summed  aggregates of all axes combined.  Presumably, this is easy for GGIR Part 2 to ingest, as metrics are usually represented as a single column.  On page 5 of your [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8301210/pdf/nihms-1659300.pdf), you mention that single-axis data are useful to investigate direction and orientation.  Our application can easily provide single-axis results as three columns, rather than an aggregated single-column.  These three columns could then be summed for an aggregated total movement.  @vincentvanhees:  Will this play nicely with GGIR?  It may complicate the input of GGIR Part 2, but it also may provide a significant benefit to users.

Thanks for continuing this discussion.  I look forward to your responses.

**Reply by vincentvanhees** [reply]
Hi Zim, GGIR already facilitates a variety of axis-specific acceleration metrics, so that should not be a problem. 
I have not worked on this since my previous comment in 2021 as this exploration was and is unpaid work for me. However, I am still interested in finding a way to integrate MIMSunit in GGIR. Preferably in a way that MIMSunit becomes a dependency of GGIR.
